### ISSUE TITLE
Memory leak when closing a PUSH socket before the PULL

### ISSUE THREAD

**Issue by takuyozora** [nanomsg/nanomsg#811]
### ISSUE

When a NN_PUSH socket bind a NN_PULL, there is a memory leak if you close the NN_PULL before the NN_PUSH.

Valgrind trace :
```c
10: ==9794== 12,104 bytes in 1 blocks are definitely lost in loss record 22 of 22
10: ==9794==    at 0x4C29BBE: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
10: ==9794==    by 0x5782562: nn_alloc_ (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x5793FA5: nn_msgqueue_init (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x579461F: nn_sinproc_init (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x57932AD: nn_cinproc_create (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x5793AD1: nn_inproc_connect (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x5776869: nn_ep_init (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x577B111: nn_sock_add_ep (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x5778FE2: nn_global_create_ep (in /usr/lib/libnanomsg.so.0969988-dirty)
10: ==9794==    by 0x5777F94: nn_connect (in /usr/lib/libnanomsg.so.0969988-dirty)
```
### REPRODUCE
```c
int sock_pull = nn_socket (AF_SP, NN_PULL);
int sock_push = nn_socket (AF_SP, NN_PUSH);
char *bind = "inproc://test";

/* Bind in socket */
nn_bind(sock_pull, bind);

/* Connect out socket */
nn_connect (sock_push, bind);


/* Free memory */
nn_close(sock_pull);
nn_close(sock_push);
```
But if you just close the push side before the pull, there is no memory leak.
```c
/* Valid : free memory */
nn_close(sock_push);
nn_close(sock_pull);
```
### DETAILS
Compile with : 
```bash
libnanomsg.so.5.0.0 => /usr/lib/libnanomsg.so.5.0.0 (0x00007f810b2c5000)
```
On archlinux x86_64 with the `community/nanomsg` package.

**Comment by gdamore** [comment]
Can you please update to nanomsg 1.0 and retest?

**Comment by takuyozora** [comment]
I really think that it's the 1.0 version : https://www.archlinux.org/packages/community/x86_64/nanomsg/

**Comment by gdamore** [comment]
Ah, ok. Will investigate and get back to you.

**Comment by gdamore** [comment]
I'm wondering if this is specific to inproc.

**Comment by gdamore** [comment]
I this will fix it:

```
commit 25b589835005b1a32f51784edf45203cca242465
Author: Garrett D'Amore <garrett@damore.org>
Date:   Wed Oct 5 21:27:14 2016 -0700

    fixes #811 Memory leak when closing a PUSH socket before the PULL

diff --git a/src/transports/inproc/cinproc.c b/src/transports/inproc/cinproc.c
index 997588b..53d2636 100644
--- a/src/transports/inproc/cinproc.c
+++ b/src/transports/inproc/cinproc.c
@@ -228,6 +228,7 @@ static void nn_cinproc_handler (struct nn_fsm *self, int src, int type,
                 nn_epbase_stat_increment (&cinproc->item.epbase,
                     NN_STAT_INPROGRESS_CONNECTIONS, 1);

+                nn_sinproc_term (&cinproc->sinproc);
                 nn_sinproc_init (&cinproc->sinproc, NN_CINPROC_SRC_SINPROC,
                     &cinproc->item.epbase, &cinproc->fsm);
                 return;
```

**Comment by gdamore** [comment]
Ok that was completley wrong.  But I've totally refactored this.  I'd appreciate a re-test using the "bug811" branch, or this PR:

https://github.com/nanomsg/nanomsg/pull/819

**Comment by takuyozora** [comment]
Ok for this one ! Thank you.
```c
10: ==15945== HEAP SUMMARY:
10: ==15945==     in use at exit: 1,144 bytes in 21 blocks
10: ==15945==   total heap usage: 286 allocs, 265 frees, 83,078 bytes allocated
10: ==15945== 
10: ==15945== LEAK SUMMARY:
10: ==15945==    definitely lost: 0 bytes in 0 blocks
10: ==15945==    indirectly lost: 0 bytes in 0 blocks
10: ==15945==      possibly lost: 0 bytes in 0 blocks
10: ==15945==    still reachable: 1,144 bytes in 21 blocks
10: ==15945==         suppressed: 0 bytes in 0 blocks
```
Have you fix the #774 ? (Yes, I hate memory leaks)

**Comment by gdamore** [comment]
Fixed in 4459720e0c413fd856eb4960c6bce86ea59dbed3
